### PR TITLE
thermalctld: Add support for fans on non-CPU modules

### PR DESCRIPTION
**Pull request by patrickmacarthur**
#### Description
This adds support to the `show platform fans` command to show fans that are on modules.

#### Motivation and Context
In the current Arista chassis model, the chassis fans are returned by `Module.get_all_fans()` as opposed to `FanDrawer.get_all_fans()`, which currently thermalctld makes no provision for. This change allows fans that are under the modules to be listed in the command output.

#### How Has This Been Tested?

This has been tested internally on a chassis, and the fan output now includes all fans on the chassis as opposed to just PSU fans:

```
admin@cmp206:~$ show platform fan
  Drawer    LED     FAN    Speed    Direction    Presence    Status          Timestamp
--------  -----  ------  -------  -----------  ----------  --------  -----------------
     N/A    off  fan0/1      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan0/2      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    red  fan0/3      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    red  fan0/4      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    red  fan0/5      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    red  fan0/6      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    red  fan0/7      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    red  fan0/8      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/1      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/2      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/3      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/4      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/5      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/6      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/7      29%      exhaust     Present        OK  20241030 15:54:08
     N/A    off  fan1/8      29%      exhaust     Present        OK  20241030 15:54:08
....
     N/A    off  psu2/1      49%       intake     Present        OK  20241030 15:54:08
     N/A    off  psu4/1      44%       intake     Present        OK  20241030 15:54:08
     N/A    off  psu6/1      44%       intake     Present        OK  20241030 15:54:08
     N/A    off  psu8/1      46%       intake     Present        OK  20241030 15:54:09
```

#### Additional Information (Optional)
Platform library support change https://github.com/sonic-net/sonic-buildimage/pull/20929 should be merged before this change.

### CODE REVIEW

**Comment by abdosi** [comment]
@bmridul : can you review this.

**Review by gechiang — APPROVED** [review]
LGTM

**Comment by gechiang** [comment]
Just want to higlight that this change has a dependency of this PR mentioned by Author in the PR description:
**Platform library support change https://github.com/sonic-net/sonic-buildimage/pull/20603 should be merged before this change.**

**Comment by rlhui** [comment]
> Just want to higlight that this change has a dependency of this PR mentioned by Author in the PR description: **Platform library support change [sonic-net/sonic-buildimage#20603](https://github.com/sonic-net/sonic-buildimage/pull/20603) should be merged before this change.**


> @bmridul and @mlok-nokia can you help check this . does this need sonic change or we should push this into platform implementation.

@bmridul , @mlok-nokia ping again

**Comment by spilkey-cisco** [comment]
I like this change! But at the same time, I think support should be added to include fans from all possible locations, not just add `Module`s. Similar to `Module`, fans may exist directly on the chassis, and not in a fan drawer. Can a type and handling for that be added in this PR as well?

**Comment by patrickmacarthur** [comment]
> I like this change! But at the same time, I think support should be added to include fans from all possible locations, not just add `Module`s. Similar to `Module`, fans may exist directly on the chassis, and not in a fan drawer. Can a type and handling for that be added in this PR as well?

This already exists today; any fan in a fan drawer on the chassis is already displayed. This PR is just adding missing output for fans that are attached to modules.

**Comment by spilkey-cisco** [comment]
> > I like this change! But at the same time, I think support should be added to include fans from all possible locations, not just add `Module`s. Similar to `Module`, fans may exist directly on the chassis, and not in a fan drawer. Can a type and handling for that be added in this PR as well?
> 
> This already exists today; any fan in a fan drawer on the chassis is already displayed. This PR is just adding missing output for fans that are attached to modules.

I'm referring specifically to fans not in a physical fan drawer (adding `CHASSIS` as a type in addition to `DRAWER`). Today, vendors must configure some notion of a 'logical' fan drawer to house fans connected directly to the chassis (not in a physical fan drawer), essentially treating the chassis itself as a fan drawer. This could be further confused by a chassis that could have both physical fan drawers with fans, and fans directly connected to the chassis without fan drawers. This feels like an unnecessary limitation, but perhaps there is some reason I'm missing as to why this should not be done?

File: sonic-thermalctld/scripts/thermalctld [hunk 1]
```diff
@@ -228,16 +234,25 @@ class FanUpdater(logger.Logger):
                 if self.task_stopping_event.is_set():
                     return
                 try:
-                    self._refresh_fan_status(drawer, drawer_index, fan, fan_index)
+                    self._refresh_fan_status(drawer, drawer_index, fan, fan_index, FanType.DRAWER)
                 except Exception as e:
                     self.log_warning('Failed to update fan status - {}'.format(repr(e)))
 
+        for module_index, module in enumerate(self.chassis.get_all_modules()):
```
Comment (spilkey-cisco) [inline sonic-thermalctld/scripts/thermalctld:241]: Thermals first check `is_chassis_system` before looping over `module` details, should that be done here as well? Or it's perhaps an unnecessary check there? https://github.com/sonic-net/sonic-platform-daemons/blob/master/sonic-thermalctld/scripts/thermalctld#L599 That also looks at PSUs connected to the modules, which may themselves have fans.

The PR title describes this as only for non-CPU modules, but it looks like that is not actually a limitation; won't this get all fans in any module where they are configured? Not necessarily a bad thing, but worth noting.

**Comment by mlok-nokia** [comment]
> > Just want to higlight that this change has a dependency of this PR mentioned by Author in the PR description: **Platform library support change [sonic-net/sonic-buildimage#20603](https://github.com/sonic-net/sonic-buildimage/pull/20603) should be merged before this change.**
> 
> > @bmridul and @mlok-nokia can you help check this . does this need sonic change or we should push this into platform implementation.
> 
> @bmridul , @mlok-nokia ping again

PR  

> > Just want to higlight that this change has a dependency of this PR mentioned by Author in the PR description: **Platform library support change [sonic-net/sonic-buildimage#20603](https://github.com/sonic-net/sonic-buildimage/pull/20603) should be merged before this change.**
> 
> > @bmridul and @mlok-nokia can you help check this . does this need sonic change or we should push this into platform implementation.
> 
> @bmridul , @mlok-nokia ping again

This 2 PR can be pushed independently.  PR20603 is Arista's platform specified code.

**Comment by patrickmacarthur** [comment]
> > > I like this change! But at the same time, I think support should be added to include fans from all possible locations, not just add `Module`s. Similar to `Module`, fans may exist directly on the chassis, and not in a fan drawer. Can a type and handling for that be added in this PR as well?
> > 
> > 
> > This already exists today; any fan in a fan drawer on the chassis is already displayed. This PR is just adding missing output for fans that are attached to modules.
> 
> I'm referring specifically to fans not in a physical fan drawer (adding `CHASSIS` as a type in addition to `DRAWER`). Today, vendors must configure some notion of a 'logical' fan drawer to house fans connected directly to the chassis (not in a physical fan drawer), essentially treating the chassis itself as a fan drawer. This could be further confused by a chassis that could have both physical fan drawers with fans, and fans directly connected to the chassis without fan drawers. This feels like an unnecessary limitation, but perhaps there is some reason I'm missing as to why this should not be done?

Discussed during chassis meeting, this should be implemented in a follow-up PR, as it widens the scope to affect not just chassis but also fixed systems.

**Review by assrinivasan — APPROVED** [review]

**Comment by bingwang-ms** [comment]
@patrickmacarthur Can you please confirm this change will not break the CLI in single ASIC platform?

**Pull request merged**